### DISCUSSION TITLE
Linux build of Tulip Desktop

### DISCUSSION THREAD

**Discussion by bwhitman** [shorepine/tulipcc#121] [Announcements]
I got this going recently. Wasn't too bad. There's a lot of testing to do. MIDI doesn't work yet. But it builds and runs fine on my Linux VM. I've updated the desktop instructions on how to build. 

<img width="1031" alt="image" src="https://github.com/bwhitman/tulipcc/assets/76612/e74da2f1-664f-43e1-8b0d-0a4eaced43f6">

**Comment by jacquesgroenewald** [comment]
Thanks for the great work on Tulip - 

Trying to follow the instructions on the Linux desktop build

Do I need to download anything? I am sure I am missing something obvious
(I am on a recent release of Ubuntu)

I ran: sudo apt install libsdl2-dev libffi-dev. This worked fine.

Now do I build it? Or how can I simply run it? 

folder tulip/linux does not exist

Thanks in advance

**Comment by bwhitman** [comment]
You first need to clone this repository, like 

```
git clone https://github.com/bwhitman/tulipcc.git
cd tulipcc/tulip/linux/
./build.sh
./dev/tulip # to run it
```

**Reply by jacquesgroenewald** [reply]
Thanks Brian!

**Reply by diflouroethane** [reply]
the `./dev/tulip` does not work, built it and everything, has this error: `./dev/tulip: No such file or directory
`

**Reply by bwhitman** [reply]
If you built it, that file should definitely exist. Can you post all of your commands their output here for building?

**Reply by diflouroethane** [reply]
i think I had an error with the build, trying something

**Reply by diflouroethane** [reply]
the last thing the build said was 

waiting for unfinished jobs....

**Reply by diflouroethane** [reply]
lol for some reason it can't find the ffi or SDL header files

**Reply by diflouroethane** [reply]
^ when I run the biuld again it can't find the ffi or SDL header files. it says ther is no such file or directory.

**Reply by diflouroethane** [reply]
i will worry about my fedora problems later, and do something else for now.

**Reply by bwhitman** [reply]
Have you run the yum commands from here: https://github.com/shorepine/tulipcc/blob/main/docs/tulip_desktop.md

`sudo yum install SDL2-devel libffi-devel`

**Reply by diflouroethane** [reply]
yes. I did, and it said something like `Package/capability 'SDL2-devel' is already requested` and exits.

**Reply by bwhitman** [reply]
Hm , yeah, this feels like a system level issue. I don't think this is related to Tulip, and I don't know enough about Fedora package management to help, sorry!

**Comment by diflouroethane** [comment]
ok, will just hope for a real one for birthday, thanks tho

**Reply by znmeb** [reply]
I got it to build on Fedora but the menu items in the lower right corner segfaulted. Do I need more dependencies installed?

**Reply by andypiper** [reply]
I'm seeing the same thing on Fedora. Also, there is no filesystem (`ls` shows a blank response, `cd('/sys')` works but does not contain the examples, etc)

**Reply by bwhitman** [reply]
`ls` should show a blank list on startup -- that's your home folder, with presumably nothing in it yet.
`cd /sys` on Desktop will cd to your computers `/sys` folder if you have one. On Tulip Desktop, our `/sys` is actually `../sys` (which should have `ex` and `im` in it. You can use `tulip.root_dir()+'sys'` for a way to do this that works on both hardware and Desktop.

**Comment by znmeb** [comment]
I have the build working in both Fedora40 and Debian 12 "bookworm" `Distrobox` containers. The Debian one works fine - "Voices" starts up, although I haven't tried to get sound out of it yet. The Fedora 40 one segfaults on "Voices" and a few of the others - I didn't try them all.

I can run in just about any conceivable `Distrobox` container, so I don't need Fedora to work. Let me know if you want me to troubleshoot it.
